File: frontend/src/types/discarded_tile.ts
/**
 * Discarded Tile type
 *
 * Backend correspondence: src/mahjong/models/discarded_tile.py::DiscardedTile
 *
 * 记录打出的牌及其打牌者信息（用于UI动画和显示）
 */

import type { Tile } from './tile';
import { createTile } from './tile';

/**
 * Discarded tile with player information
 */
export interface DiscardedTile {
  tile: Tile;         // 打出的麻将牌
  playerId: string;   // 打牌者ID（"human", "ai_1", "ai_2", "ai_3"）
  turnIndex: number;  // 第几张打出的牌（从0开始，用于排序和动画）
}

/**
 * Utility function: Create DiscardedTile from backend data
 */
export function createDiscardedTile(data: any): DiscardedTile {
  return {
    tile: createTile(data.tile),
    playerId: data.playerId,
    turnIndex: data.turnIndex
  };
}

File: src/mahjong/models/discarded_tile.py
"""
Discarded Tile Model

记录打出的牌及其打牌者信息
"""

from dataclasses import dataclass
from mahjong.models.tile import Tile


@dataclass(frozen=True)
class DiscardedTile:
    """记录打出的牌及其打牌者

    Attributes:
        tile: 打出的麻将牌
        player_id: 打牌者ID（"human", "ai_1", "ai_2", "ai_3"）
        turn_index: 第几张打出的牌（从0开始，用于排序和动画）
    """
    tile: Tile
    player_id: str
    turn_index: int

    def to_dict(self) -> dict:
        """序列化为字典（用于API响应）"""
        return {
            "tile": {
                "suit": self.tile.suit.name,
                "rank": self.tile.rank
            },
            "playerId": self.player_id,
            "turnIndex": self.turn_index
        }
